OneCommunityGlobal/HighestGoodNetworkApp
# Fix Badges Select Featured and Edit Personal Max Functions

# Description
https://www.loom.com/share/737001d350ca4b79ab9950a158df7b3d?sid=177c042d-ba44-4cf1-8daa-a9b712537181
1. In "Select Featured", select the featured badges (at most five) and save changes.The web will crash.
2. In "Assign Badges", the checkbox does not check the badges which user already have.
3. For "New Max - Personal Record Award" badge, it cannot be modified in "Select Featured".

## Related PRS (if any):
This frontend PR is related to backend development branch.
…

## Main changes explained:
- Update returnUpdatedBadgesCollection function in `badgeManagement.js` and `AssignBadgePopup.jsx` to make sure the functionality of checkbox can add / reassign the badges in selectedBadges.
- Update useEffect in `BadgeReport.jsx` to avoid memory leak after Save Changes in Select Featured.
- Update ` BadgeImage.jsx` to adjust the value for "New Max - Personal Record Award" badge.
- Update ` AssignBadgePopup.jsx` to rename selectedBadges to existBadges to avoid use same as Redux one and simulate the hook again and again
- Update related test cases to adopt to new logic

## How to test:
1. check into current branch
2. do `npm install` and `npm run start:local` to run this PR locally
3. Clear site data/cache
5. Log as owner / admin user
6. Go to Profile Page>Select Featured / Assign Badges
7. Verify
        (a.) When open the "Assign Badges" popup, see the checkbox check the existing badges.
        (b.) Test the functionality: Select the existing badge then its amount will +1. Select the new badge will show 1 except the "New Max - Personal Record Award" badge.
        (c.) When select "New Max - Personal Record Award" badge, it will first show the "personalBestMaxHrs" in userProfile from backend(you can check that in MongoDB). And, if you adjust the Count value in "Select Featured" and larger than "personalBestMaxHrs", it will show the Count value.
        (d.) Select the featured badges (at most five) and save changes and see whether the web will crash or not.

## Screenshots or videos of changes:
https://www.loom.com/share/e7ab95c9c2e44191847f441cd3c5622e?sid=546d9147-e6b7-4fef-a7f5-f694da367ddc
